nncf/experimental/common/pruning/propagation_data.py

```diff
@@ -0,0 +1,221 @@
+    head. In that case, the propagation block is encoded by size=N and offset=0.
+
+    size = 6 (head size)
+    offset = 0
+    number of all channels = 18
+
+    all channels           ------------------
+    block #1, mask: 011    xxxxxx------------
+    block #2, mask: 101    ------xxxxxx------
+    block #3, mask: 110    ------------xxxxxx
+    mask: 100              ------xxxxxxxxxxxx
+
+    3) Or potentially, one can remove the same dimension in each Transformer's head. The propagation block would be as
+    follows: size=1 and offset=N. It means that block is formed by taking the `size` number of sequent channels, then
+    skipping the `offset` number of sequent channels and repeating the procedure for the rest channels. For that
+    particular case, if total number of channels is 3N, the size of block will be equal to 3.
+    Block will contain 6 elements, if size=2 offset=N with the same size of pruning mask.
+
+    size = 1
+    offset = 6 (head size)
+    number of all channels = 18
+
+    all channels              ------------------
+    block #1, mask: 011111    x-----x-----x-----
+    block #2, mask: 101111    -x-----x-----x----
+    ...
+    block #6, mask: 111110    -----x-----x-----x
+    mask 110000               --xxxx--xxxx--xxxx
+    """
+    def __init__(self,
+                 producer: MaskProducer,
+                 size: int = 1,
+                 offset: int = 0,
+                 pruning_dimension: int = 0) -> None:
+        """
+
+        :param producer: descriptor of the producer
+        :param size: number of sequent channels.
+        :param offset: when not equal to 0, block is formed by taking `size` number of sequent channels,
+        then skipping `offset` number of sequent channels and repeating the procedure for the rest of channels.
+        :param pruning_dimension: axis number from 0 to N-1 in weights along which the dimension block defines pruning
+        structure. N is total number of dimensions.
+        """
+        self.size = size
+        self.offset = offset
+        self.pruning_dimension = pruning_dimension
+        self._producer = producer
+
+    def __eq__(self, other) -> bool:
+        return self.pruning_dimension == other.pruning_dimension and \
+            self.size == other.size and \
+            self.offset == other.offset and \
+            self._producer.id == other._producer.id
+
+    def __str__(self) -> str:
+        return f"S:{self.size}__O:{self.offset}__ID:{self._producer.id}"
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+
+class PropagationGroup:
+    """
+    Defines a group of propagation blocks and links it with the list of children groups.
+    The group is initialized on producers of pruning and is propagated until consumers within PropagationMask.
+    """
+
+    def __init__(self, blocks: List[PropagationBlock]) -> None:
+        self._blocks = blocks
+        self._children: List['PropagationGroup'] = []
+        self.is_invalid = False
```

Comment:
Can we define `is_invalid` attr as property?

Reply:
Attribute `is_invalid` should be set only by `invalidate` call, so I suggest to make `is_valid` access rule read only